hi-ml-multimodal/src/health_multimodal/image/model/encoder.py

```diff
@@ -0,0 +1,173 @@
+#  -------------------------------------------------------------------------------------------
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+#  -------------------------------------------------------------------------------------------
+
+from __future__ import annotations
+
+from contextlib import contextmanager
+from typing import Any, Generator, Optional, Sequence, Tuple, Union
+
+import torch
+import torch.nn as nn
+from health_multimodal.common.device import get_module_device
+from timm.models.layers import trunc_normal_
+
+from .resnet import resnet18, resnet50
+from .transformer import VisionTransformerPooler
+from .types import ImageEncoderType
+
+TypeImageEncoder = Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
+
+
+class ImageEncoder(nn.Module):
+    """Image encoder trunk module for the ``ImageModel`` class.
+
+    :param img_model_type: Type of image model to use: either ``"resnet18"`` or ``"resnet50"``.
+    """
+
+    def __init__(self, img_model_type: str):
+        super().__init__()
+        self.img_model_type = img_model_type
+        self.encoder = self._create_encoder()
+
+    def _create_encoder(self, **kwargs: Any) -> nn.Module:
+        if self.img_model_type in [ImageEncoderType.RESNET18, ImageEncoderType.RESNET18_MULTI_IMAGE]:
+            encoder_class = resnet18
+        elif self.img_model_type in [ImageEncoderType.RESNET50, ImageEncoderType.RESNET50_MULTI_IMAGE]:
+            encoder_class = resnet50
+        else:
+            supported = ImageEncoderType.get_members(multi_image_encoders_only=False)
+            raise NotImplementedError(f"Image model type \"{self.img_model_type}\" must be in {supported}")
+
+        encoder = encoder_class(pretrained=True, **kwargs)
+
+        return encoder
+
+    def forward(self,
+                current_image: torch.Tensor,
+                return_patch_embeddings: bool = False) -> TypeImageEncoder:
+        """Image encoder forward pass."""
+
+        patch_emb = self.encoder(current_image)
+        avg_pooled_emb = torch.flatten(torch.nn.functional.adaptive_avg_pool2d(patch_emb, (1, 1)), 1)
+        if return_patch_embeddings:
+            return patch_emb, avg_pooled_emb
+
+        return avg_pooled_emb
+
+    def reload_encoder_with_dilation(self, replace_stride_with_dilation: Optional[Sequence[bool]] = None) -> None:
+        """Workaround for enabling dilated convolutions after model initialization.
+
+        :param replace_stride_with_dilation: for each layer to replace the 2x2 stride with a dilated convolution
```

Comment:
I guess it's not really for each layer as the sequence in line 69 has length 3 only.